File: ObjC/IAPController.m
//
//  IAPController.m
//  Pods
//
//  Created by materik on 03/01/16.
//
//

#import "IAPController.h"

NSString *const IAPControllerFetchedNotification =
    @"IAPControllerFetchedNotification";
NSString *const IAPControllerPurchasedNotification =
    @"IAPControllerPurchasedNotification";
NSString *const IAPControllerFailedNotification =
    @"IAPControllerFailedNotification";
NSString *const IAPControllerRestoredNotification =
    @"IAPControllerRestoredNotification";

static IAPController *_sharedInstance;

@interface IAPController () <SKProductsRequestDelegate,
                             SKPaymentTransactionObserver>

@property(nonatomic, strong) NSArray<NSString *> *productIds;

@end

@implementation IAPController

@synthesize products = _products;

+ (IAPController *)sharedInstance {
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    _sharedInstance = [[IAPController alloc] init];
  });
  return _sharedInstance;
}

- (instancetype)init {
  self = [super init];
  if (self) {
    [self retrieveProductIdsFromPlist];
    [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
  }
  return self;
}

- (void)retrieveProductIdsFromPlist {
  NSURL *url = [[NSBundle mainBundle] URLForResource:@"IAPControllerProductIds"
                                       withExtension:@"plist"];
  self.productIds = [NSArray arrayWithContentsOfURL:url];
}

#pragma mark - Action

- (void)fetchProducts {
  SKProductsRequest *request = [[SKProductsRequest alloc]
      initWithProductIdentifiers:[NSSet setWithObjects:self.productIds, nil]];
  [request setDelegate:self];
  [request start];
}

- (void)restore {
  [[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
}

#pragma mark - SKPaymentTransactionObserver

- (void)productsRequest:(SKProductsRequest *)request
     didReceiveResponse:(SKProductsResponse *)response {
  _products = response.products;
  [[NSNotificationCenter defaultCenter]
      postNotificationName:IAPControllerFetchedNotification
                    object:nil];
}

- (void)paymentQueue:(SKPaymentQueue *)queue
 updatedTransactions:(NSArray<SKPaymentTransaction *> *)transactions {
  for (SKPaymentTransaction *transaction in transactions) {
    switch (transaction.transactionState) {
    case SKPaymentTransactionStatePurchased:
      [self purchasedTransaction:transaction];
      break;
    case SKPaymentTransactionStateFailed:
      [self failedTransaction:transaction error:nil];
      break;
    case SKPaymentTransactionStateRestored:
      [self restoreTransaction:transaction];
      break;
    default:
      break;
    }
  }
}

- (void)paymentQueue:(SKPaymentQueue *)queue
    restoreCompletedTransactionsFailedWithError:(NSError *)error {
  [self failedTransaction:nil error:error];
}

- (void)paymentQueueRestoreCompletedTransactionsFinished:
    (SKPaymentQueue *)queue {
  [[NSNotificationCenter defaultCenter]
      postNotificationName:IAPControllerRestoredNotification
                    object:nil];
}

#pragma mark - Transactions

- (void)finishTransaction:(SKPaymentTransaction *)transaction {
  if (transaction) {
    [[SKPaymentQueue defaultQueue] finishTransaction:transaction];
  }
}

- (void)restoreTransaction:(SKPaymentTransaction *)transaction {
  [self finishTransaction:transaction];
  [[NSNotificationCenter defaultCenter]
      postNotificationName:IAPControllerPurchasedNotification
                    object:nil];
}

- (void)failedTransaction:(SKPaymentTransaction *)transaction
                    error:(NSError *)error {
  [self finishTransaction:transaction];
  error = error ?: transaction.error;
  if (error) {
    if (error.code != SKErrorPaymentCancelled) {
      [[NSNotificationCenter defaultCenter]
          postNotificationName:IAPControllerFailedNotification
                        object:error];
    }
  }
}

- (void)purchasedTransaction:(SKPaymentTransaction *)transaction {
  [self finishTransaction:transaction];
  [[NSNotificationCenter defaultCenter]
      postNotificationName:IAPControllerPurchasedNotification
                    object:nil];
}

@end

File: Swift/IAPController.swift
//
//  IAPController.swift
//  XWorkout
//
//  Created by materik on 19/07/15.
//  Copyright (c) 2015 materik. All rights reserved.
//

import Foundation
import StoreKit

public let IAPControllerFetchedNotification = "IAPControllerFetchedNotification"
public let IAPControllerPurchasedNotification = "IAPControllerPurchasedNotification"
public let IAPControllerFailedNotification = "IAPControllerFailedNotification"
public let IAPControllerRestoredNotification = "IAPControllerRestoredNotification"

open class IAPController: NSObject, SKProductsRequestDelegate, SKPaymentTransactionObserver {
    
    // MARK: Properties
    
    open var products: [SKProduct]?
    var productIds: [String] = []
    
    // MARK: Singleton
    
    open static var sharedInstance: IAPController = {
        return IAPController()
    }()
    
    // MARK: Init
    
    public override init() {
        super.init()
        self.retrieveProductIdsFromPlist()
        SKPaymentQueue.default().add(self)
    }
    
    private func retrieveProductIdsFromPlist() {
        let url = Bundle.main.url(forResource: "IAPControllerProductIds", withExtension: "plist")!
        self.productIds = NSArray(contentsOf: url) as! [String]
    }
    
    // MARK: Action
    
    open func fetchProducts() {
        let request = SKProductsRequest(productIdentifiers: Set(self.productIds))
        request.delegate = self
        request.start()
    }
    
    open func restore() {
        SKPaymentQueue.default().restoreCompletedTransactions()
    }
    
    // MARK: SKPaymentTransactionObserver

    open func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
        self.products = response.products
        NotificationCenter.default.post(name: Notification.Name(rawValue: IAPControllerFetchedNotification), object: nil)
    }
    
    open func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
        for transaction in transactions {
            switch transaction.transactionState {
            case .purchased:
                self.purchasedTransaction(transaction: transaction)
                break
            case .failed:
                self.failedTransaction(transaction: transaction)
                break
            case .restored:
                self.restoreTransaction(transaction: transaction)
                break
            default:
                break
            }
        }
    }
    
    open func paymentQueue(_ queue: SKPaymentQueue, restoreCompletedTransactionsFailedWithError error: Error) {
        self.failedTransaction(transaction: nil, error: error)
    }
    
    open func paymentQueueRestoreCompletedTransactionsFinished(_ queue: SKPaymentQueue) {
        NotificationCenter.default.post(name: Notification.Name(rawValue: IAPControllerRestoredNotification), object: nil)
    }
    
    // MARK: Transaction
    
    func finishTransaction(transaction: SKPaymentTransaction? = nil) {
        if let transaction = transaction {
            SKPaymentQueue.default().finishTransaction(transaction)
        }
    }
    
    func restoreTransaction(transaction: SKPaymentTransaction? = nil) {
        self.finishTransaction(transaction: transaction)
        NotificationCenter.default.post(name: Notification.Name(rawValue: IAPControllerPurchasedNotification), object: nil)
    }
    
    func failedTransaction(transaction: SKPaymentTransaction? = nil, error: Error? = nil) {
        self.finishTransaction(transaction: transaction)
        if let error = error ?? transaction?.error {
            if error._code != SKError.paymentCancelled.rawValue {
                NotificationCenter.default.post(name: Notification.Name(rawValue: IAPControllerFailedNotification), object: error)
            }
        }
    }
    
    func purchasedTransaction(transaction: SKPaymentTransaction? = nil) {
        self.finishTransaction(transaction: transaction)
        NotificationCenter.default.post(name: Notification.Name(rawValue: IAPControllerPurchasedNotification), object: nil)
    }
    
}

